test/com/amplitude/api/SessionTest.java

```diff
@@ -992,4 +992,60 @@ public void testOutOfSessionIdentifyDoesNotTriggerNewSession() throws JSONExcept
             identifies.getJSONObject(0).getJSONObject("user_properties"), expected
         ));
     }
+
+    @Test
+    public void testSetUserIdAndStartNewSessionWithTracking() {
+        amplitude.trackSessionEvents(true);
+
+        long timestamp = System.currentTimeMillis();
+        amplitude.logEventAsync("test", null, null, null, null, timestamp, false);
+        Shadows.shadowOf(amplitude.logThread.getLooper()).runToEndOfTasks();
+
+        // trackSessions is true, start_session event is added
+        assertEquals(getUnsentEventCount(), 2);
+
+        // set user id and validate session ended and new session started
+        amplitude.setUserId("test_new_user", true);
```

Comment:
Maybe test a call with false to make sure there are no extra events being generated? And try a call with session events off?

Reply:
good idea. done